crates/rerun/src/clap.rs

```diff
@@ -58,51 +62,78 @@ pub struct RerunArgs {
 }
 
 impl RerunArgs {
-    /// Run common Rerun script setup actions. Connect to the viewer if necessary.
+    /// Set up Rerun, and run the given code with a [`Session`] object
+    /// that can be used to log data.
     ///
-    /// Returns `true` if you should call `session.spawn`.
-    pub fn on_startup(&self, session: &mut Session) -> anyhow::Result<bool> {
-        match self.to_behavior() {
-            RerunBehavior::Connect(addr) => session.connect(addr),
+    /// Logging will be controlled by the `RERUN` environment variable,
+    /// or the `default_enabled` argument if the environment variable is not set.
+    #[track_caller] // track_caller so that we can see if we are being called from an official example.
+    pub fn run(
+        &self,
+        application_id: &str,
+        default_enabled: bool,
+        run: impl FnOnce(Session) + Send + 'static,
+    ) -> anyhow::Result<()> {
+        let (rerun_enabled, recording_info) = crate::SessionBuilder::new(application_id)
+            .default_enabled(default_enabled)
+            .finalize();
+
+        if !rerun_enabled {
+            run(Session::disabled());
+            return Ok(());
+        }
+
+        let sink: Box<dyn re_sdk::sink::LogSink> = match self.to_behavior()? {
+            RerunBehavior::Connect(addr) => Box::new(crate::sink::TcpSink::new(addr)),
 
-            RerunBehavior::Save(path) => session.save(path)?,
+            RerunBehavior::Save(path) => Box::new(crate::sink::FileSink::new(path)?),
 
             #[cfg(feature = "web_viewer")]
             RerunBehavior::Serve => {
-                crate::serve_web_viewer(session, true);
+                let open_browser = true;
+                crate::web_viewer::new_sink(open_browser)
             }
 
-            RerunBehavior::Spawn => return Ok(true),
-        }
+            #[cfg(feature = "native_viewer")]
+            RerunBehavior::Spawn => {
+                crate::native_viewer::spawn(recording_info, run)?;
+                return Ok(());
+            }
+        };
 
-        Ok(false)
-    }
+        let session = Session::new(recording_info, sink);
+        run(session);
 
-    /// Run common post-actions. Sleep if serving the web viewer.
-    pub fn on_teardown(&self) {
         #[cfg(feature = "web_viewer")]
-        if self.to_behavior() == RerunBehavior::Serve {
-            eprintln!("Sleeping while serving the web viewer. Abort with Ctrl-C"); // TODO(emilk): sleep in `drop` instead?
+        if matches!(self.to_behavior(), Ok(RerunBehavior::Serve)) {
+            eprintln!("Sleeping while serving the web viewer. Abort with Ctrl-C");
             std::thread::sleep(std::time::Duration::from_secs(1_000_000));
```

Comment:
Drive-by nitpicking: 1 million seconds isn't all that long... 11 days is a long, but totally achievable session length.

Reply:
good catch!